### DISCUSSION TITLE
DateTimeSensor does not respect poke_interval when executing from PythonOperator

### DISCUSSION THREAD

**Discussion by kolfild26** [apache/airflow#33682] [General]
### Apache Airflow version

Other Airflow 2 version (please specify below)

### **v.2.6.3**

Executing the DateTimeSensor with `reschedule` mode from inside the PythonOperator leads to strange behaviour.

### How to reproduce

```
def _wait(**context):
    target_time = datetime.now(timezone.utc) + timedelta(hours=5)
    wait = DateTimeSensor(
        task_id="wait",
        target_time=target_time,
        poke_interval=10,
        timeout=10000,
        mode="reschedule",
        soft_fail=False,
        )
    wait.execute(context=context)
    
wait_trigger = PythonOperator(
    task_id="wait_trigger",
    python_callable=_wait,
    )

wait_trigger 
```
What's happening ⬇️ 
1) Despite poke_interval is set as 10 seconds, it actually runs every second. See below.
2) Moreover, every time (1 second) it updates its target_time

```
[2023-08-23, 23:35:35 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:35.753491+00:00) has come
[2023-08-23, 23:35:36 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:36.823583+00:00) has come
[2023-08-23, 23:35:37 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:37.906547+00:00) has come
[2023-08-23, 23:35:39 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:38.986302+00:00) has come
[2023-08-23, 23:35:39 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:39.760245+00:00) has come
[2023-08-23, 23:35:40 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:40.823194+00:00) has come
[2023-08-23, 23:35:41 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:41.911287+00:00) has come
[2023-08-23, 23:35:43 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:42.994946+00:00) has come
[2023-08-23, 23:35:43 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:43.558043+00:00) has come
[2023-08-23, 23:35:44 ] {{date_time.py:74}} INFO - Checking if the time (2023-08-24T01:35:44.625291+00:00) has come
```

As for the tagret time, it's because i'm using now() in the example above. Wiil be no problem if it's fixed. But this is a symptom that every 1 second the scheduler re-run `_wait` function again.
I thought things should be going like:
- It executes the PythonOperator once
- There is a callabe function inside the PythonOperator which is also supposed to be running ONCE
- It creates and executes the DateTimeSensor task
- Then the DateTimeSensor works based on those parameters which was passed into it in the callable.

But in real life it executes the callable again each time it reschedules the task which is executing in the callable itself.
And even assuming it's an expected behaviour (is it?), why `poke_interval` is not taking into account is still unclear.

p.s All this happens when the `reschedule` mode is set

### Operating System

rhel

### Deployment

Virtualenv installation

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by Taragolis** [comment]
You should not manually run `execute` method of operators, this method intend to use by Airflow itself.

Between initialise operator and execute it a lot of internal things happen which required for proper work Airflow itself: scheduling, execution, monitoring, and etc, but if you run `execute` in your code, you just skip this steps and behaviour might be unpredictable.

For example, sensors in rescheduling mode just execute once and if result False send signal to scheduler for reschedule it.

I would recommend write your own operator/sensor, you could create it from scratch or overwrite one of existed.

Some useful links for reference:
- [Creating a custom Operator](https://airflow.apache.org/docs/apache-airflow/stable/howto/custom-operator.html)
- [Public Interface of Airflow](https://airflow.apache.org/docs/apache-airflow/stable/public-airflow-interface.html)
- [How-to Guides](https://airflow.apache.org/docs/apache-airflow/stable/howto/index.html)

**Comment by kolfild26** [comment]
Actually what I was trying to acheive is to set up target_time, poke_interval and timeout in the dargun config on the fly, then access them from `{{ dag_run.conf }}` and pass to the correspondent task arguments . The problem is that `target_time` is templated but `poke_interval` and `timeout` are not.
Definitely, no other way-out but to create a custom one.

**Reply by potiuk** [reply]
There are two ways out:
1) Create a custom operator and set templated fields as you see fit (and tha's perfectly OK - customising operators when built-in ones are not enough is first-class citizen in Airlfow). And it is far less code and easier (and absolutely standard way of extending Airflow capabilities).
2) Contribute the change back to airflow

But definitely what you tried to do is neither standard, nor going to work :)

**Comment by kolfild26** [comment]
@potiuk 
Ok, let me ask more. I previously found some threads ([[1](https://stackoverflow.com/questions/74959020/airflow-python-question-calling-operator-within-python-operator)],[[2](https://stackoverflow.com/questions/52915677/how-to-trigger-operator-inside-python-function-using-airflow)],[[3](https://stackoverflow.com/questions/64295326/airflow-use-an-operator-inside-a-function/64297270#64297270)]) on stackoverflow with the common question of "how to use operator inside function". All there they advise to call the execute method and so on.

So, do you mean all those advices is a bad pattern and should not be used?
If so, then we don't have any allowed way to call operator inside function, right?

**Comment by potiuk** [comment]
> @potiuk Ok, let me ask more. I previously found some threads ([[1](https://stackoverflow.com/questions/74959020/airflow-python-question-calling-operator-within-python-operator)],[[2](https://stackoverflow.com/questions/52915677/how-to-trigger-operator-inside-python-function-using-airflow)],[[3](https://stackoverflow.com/questions/64295326/airflow-use-an-operator-inside-a-function/64297270#64297270)]) on stackoverflow with the common question of "how to use operator inside function". All there they advise to call the execute method and so on.
> 
> So, do you mean all those advices is a bad pattern and should not be used? If so, then we don't have any allowed way to call operator inside function, right?

I have no time or energy to look at all the random answers people did and spend my time on analysing them just to satisfy someone's curiosity. Maybe there were different cases, maybe you misunderstood questions or answers, or maybe they were just wrong. All of those things happen. But yes. you should never need to call execute() manually - it makes no sense. Execute () is called by airflow when the time comes and airflow does a lot more than calling execute(). It is unreasonable assumption that something "more" happens when you call execute() - it's just calling a python method. If you know how python works - this is what happened. You called a method on an object. That's it. Nothing more could happen. There is no way retries, reschedules etc. or anything else happens there. DON't do it.

> If so, then we don't have any allowed way to call operator inside function, right?

You CAN call it. But don't expect anything more than calling an execute method on an object. You seem you expect more (to happen magically)
